### ISSUE TITLE
`:redundant-ignore` false positive for finding from hook

### ISSUE THREAD

**Issue by lread** [clj-kondo/clj-kondo#2433]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version**
```
$ clj-kondo --version
clj-kondo v2024.11.14
```

**macro usage**
Maybe kinda related.

**platform**
All

**problem**
On Etaoin, I noticed that clj-kondo is falsely reporting `:redundant-ignore`s.
The issue seems to be related to custom linters in clj-kondo hooks.

**repro**
I made a minimal repro: https://github.com/lread/repro-kondo-redundant-ignore-custom-linter

```
git clone https://github.com/lread/repro-kondo-redundant-ignore-custom-linter
cd repro-kondo-redundant-ignore-custom-linter
clj-kondo --lint src
```

**actual behavior**
clj-kondo reports:
```
src/foobar/baz.clj:16:3: info: Redundant ignore
src/foobar/baz.clj:21:8: error: How dare foo!
linting took 16ms, errors: 1, warnings: 0
```

**expected behavior**
I did not expect the redundant ignore.

See repro [README](https://github.com/lread/repro-kondo-redundant-ignore-custom-linter/blob/main/README.adoc) for more details.

**diagnosis**
I did a little poking around and suspect it might have something to do with hooks and `derived-data`. 

**next steps**
I'm happy to poke around further if that helps.

**Comment by borkdude** [comment]
> next steps
> I'm happy to poke around further if that helps.

More than welcome to!

**Comment by lread** [comment]
Ok, here are some notes from an initial peek (more to follow).

From a REPL, I used to [Flowstorm](https://github.com/flow-storm/flow-storm-debugger) to trace hitting my repro repo (from a `comment` block in `clj-kondo.core`):

```clojure
(-> (run!
       {;; seq of string or file
        :lint ["/home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/src"]
        :config-dir "/home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/.clj-kondo"})
      :findings)
```

At entry of [clj-kondo.impl.linters/lint-redundant-ignores](https://github.com/clj-kondo/clj-kondo/blob/82bec59f4ba3bd48cf037043c0c6fae1120ca096/src/clj_kondo/impl/linters.clj#L879-L893), `(:ignores ctx)` contains:
```clojure
{"/home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/src/foobar/baz.clj"
 {:clj
  [{:row 17,
    :col 1,
    :end-row 17,
    :end-col 22,
    :clj-kondo/ignore
    {:row 16, :col 3, :end-row 16, :end-col 43, :linters nil},
    :ignore #{:acme/foo-not-good},
    :used true}
   {:row 17,
    :col 1,
    :end-row 17,
    :end-col 22,
    :clj-kondo/ignore
    {:row 16, :col 3, :end-row 16, :end-col 43, :linters nil},
    :derived-location true,
    :ignore #{:acme/foo-not-good}}
   {:row 17,
    :col 1,
    :end-row 17,
    :end-col 22,
    :clj-kondo/ignore
    {:row 16, :col 3, :end-row 16, :end-col 43, :linters nil},
    :derived-location true,
    :ignore #{:acme/foo-not-good}}
   {:row 17,
    :col 1,
    :end-row 17,
    :end-col 22,
    :clj-kondo/ignore
    {:row 16, :col 3, :end-row 16, :end-col 43, :linters nil},
    :derived-location true,
    :ignore #{:acme/foo-not-good}}]}}
```
The last three ignores are not marked as `used` and, therefore, registered as findings.
They are later deduped into a single finding in `clj-kondo.core/run!`.

Notice that these last three ignores all have `:derived-location true`.

Some debugging tells me these four ignores were added in [clj-kondo.impl.utils/handle-ignore](https://github.com/clj-kondo/clj-kondo/blob/82bec59f4ba3bd48cf037043c0c6fae1120ca096/src/clj_kondo/impl/utils.clj#L368-L387) for the following `expr`s:

1. `(dingo foo (str (foo)))`
2. `(let [foo 42] (str foo))`
3. `let`
4. `42`

The repro code on [line 16 and 17](https://github.com/lread/repro-kondo-redundant-ignore-custom-linter/blob/e766c5ba037634e6b1310a6b5dd2456b1eca2f67/src/foobar/baz.clj#L16-L17) is:

```clojure
#_{:clj-kondo/ignore [:acme/foo-not-good]}
(dingo foo (str foo))
```

And the [repro hook code that generates the rewrite-clj node](https://github.com/lread/repro-kondo-redundant-ignore-custom-linter/blob/e766c5ba037634e6b1310a6b5dd2456b1eca2f67/.clj-kondo/org/acme/org/acme.clj#L13-L17) is:

```clojure
    (let [new-node (api/list-node
                     (list*
                       (api/token-node 'let)
                       (api/vector-node [somesym (api/token-node 42)])
                       somebody))]
```

Next up:

1. Why are the `derived-location` ignores being generated?
2. Should they be?

**Comment by borkdude** [comment]
> 1. Why are the derived-location ignores being generated?

This is done for hooks so artifically created nodes still have a location to point to in case of a warning

2. Yes, I think they should be, but it seems it has an undesired side effect with respected to unused ignores

I think the ignore should not be copied over to the artificial nodes

**Comment by borkdude** [comment]
I'm not 100% sure about the approach to solve this yet, but the issue here is that:

https://github.com/clj-kondo/clj-kondo/blob/82bec59f4ba3bd48cf037043c0c6fae1120ca096/src/clj_kondo/impl/hooks.clj#L134

we copy the metadata from the original node to any inner generated nodes by the hooks. 

- If we wouldn't copy the ignores, then any errors may still be reported for hook generated code when a user adds an ignore hint. We should probably write a test for this, if there isn't any yet.
- If we do copy the ignores (and keep doing so) perhaps we need to add to the ignore that it should already be regarded to be used so it won't be marked unused.

Perhaps an approach is to mark an ignore hint with a generated id. We could then check in the ctx ignores if it isn't already there and just not add it if it's already there.

**Comment by lread** [comment]
Thanks for the details!

Probably another simpler interesting case to cover is some non-custom linter finding from hook-generated nodes:

```clojure
;; problem, unused binding `moodog`
(dingo moodog (println "hello"))

;; unused binding suppressed
#_{:clj-kondo/ignore [:unused-binding]}
(dingo moodog (println "hello"))
```

[I've tacked this case onto my repro repo](https://github.com/lread/repro-kondo-redundant-ignore-custom-linter/blob/f18bc1be7e526a5e517d29fdbb38da4e7af08a71/src/foobar/baz.clj#L24-L31).

Here's the new results:

```
$ clj-kondo --lint src
src/foobar/baz.clj:16:3: info: Redundant ignore
src/foobar/baz.clj:21:8: error: How dare foo!
src/foobar/baz.clj:27:8: warning: unused binding moodog
src/foobar/baz.clj:30:3: info: Redundant ignore
linting took 14ms, errors: 1, warnings: 1
```
New findings:
- Line 27 is finding makes sense.
- Line 30 is finding is incorrect.

Nothing new, but a simpler use case that shows the issue is not triggered specifically by custom linters from hooks.

**Comment by lread** [comment]
> If we do copy the ignores (and keep doing so) perhaps we need to add to the ignore that it should already be regarded to be used so it won't be marked unused.

Ok, If I hack an assoc of `:used true` on `!!last-meta`:

```clojure
(defn annotate
  {:no-doc true}
  [node original-meta]
  (let [!!last-meta (volatile! (assoc original-meta :derived-location true :used true))]
```

All existing tests pass.

And my repro repo returns expected findings:
```
$ clojure -M:clj-kondo/dev \                                                                                                        
        --lint /home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/src \
        --config-dir /home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/.clj-kondo
/home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/src/foobar/baz.clj:21:8: error: How dare foo!
/home/lee/proj/oss/lread/repro-kondo-redundant-ignore-custom-linter/src/foobar/baz.clj:27:8: warning: unused binding moodog
linting took 139ms, errors: 1, warnings: 1
```

Without a complete understanding of annotating nodes generated from hooks, I cannot say whether or not this is a smart approach.

**Comment by borkdude** [comment]
I'll have a look tomorrow, thanks for the research so far, very helpful!

**Comment by lread** [comment]
Glad to be of some help!

**Comment by borkdude** [comment]
I believe https://github.com/clj-kondo/clj-kondo/commit/b03d8e82ad1a5723789fd9f94a709b984e8496f9 might fix it... Just need to add a test
